ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java

```diff
@@ -704,6 +750,55 @@ protected void deleteBlob(final String blobIdStr, final String serviceId, Future
       routerCallback.onPollReady();
     }
 
+    /**
+     * Requests for a blob to be undeleted asynchronously and invokes the {@link Callback} when the request completes.
+     * @param blobIdStr The ID of the blob that needs to be undeleted in string form
+     * @param serviceId The service ID of the service undeleting the blob. This can be null if unknown.
+     * @param futureResult A future that would contain information about whether the undelete succeeded or not,
+     *                     eventually.
+     * @param callback The {@link Callback} which will be invoked on the completion of a request.
+     */
+    protected void undeleteBlob(final String blobIdStr, final String serviceId, FutureResult<Void> futureResult,
```

Comment:
The code looks like similar to updateBlobTtl. Is it possible to unify them?

Reply:
I've tried to reuse most of them. The tricky. part is about those metrics and compete methods. I suppose there is way to reuse them but it seems not really necessary.

Reply:
> The code looks like similar to updateBlobTtl. Is it possible to unify them?

I made some new change, will this make you happy?